Refactor and simplify all our smell detector specs.

troessner opened the pull request:
Follow up to #1035 after this one got bigger and bigger now with a better branch name and a better commit message.

This is still - again - wip. Right now I'm literally going through all the specs to make them simpler and more concise (my current position is FeatureEnvy). This will take me a couple of more days.

What I did:

* Killed all the shared examples
* Introduced our Spec matcher everywhere. This and the bullet point above also have the nice side effect that now all requires at the top of a smell detector spec file look
like this:

```Ruby
require_relative '../../spec_helper'
require_lib 'reek/smells/irresponsible_module'
````

And that's it. No more references to `Examiner`, `CodeContext` and god knows what else.

* Removed silly specs. Some areas of Reek are shockingly under-speced whilst others are over-speced. If we have a threshold of 4 x it is just silly to have a spec that 3 x
does not smell. And 2 x does not smell. And 1 x does not smell. That's just bloated code.

* Removed redundant specs. Bigger smell detector specs were full of duplication.

* Removed foo / bar because I honestly can't see that anymore and replaced it with "alpha/bravo/charlie" as discussed.

* Made our syntax consistent. Introduced heredocs everywhere and removed multi line strings. Introduced newlines. And so on.

* Simplified a lot, a lot of our examples. For instance:

```Ruby
src = 'def simple(arg) args = {}; if arg then args.merge(\'a\' => \'A\') end end'
# to
src = 'def alpha(bravo); if bravo then charlie end; end'
# or
src = 'def <=>(other) Options[:sort_order].compare(self, other) end'
# to
src = 'def alfa(bravo); alfa.bravo(self); end'
```

The meaning stays the same but the example is less confusing and better readable.

* Introduced the same general examples for every spec at the beginning. Every smell detector spec now starts out like this:

```Ruby
RSpec.describe Reek::Smells::IrresponsibleModule do
  it 'reports the right values' do
    src = <<-EOS
      class Dummy
      end
    EOS

    expect(src).to reek_of(described_class,
                           lines:   [1],
                           context: 'Dummy',
                           message: 'has no descriptive comment',
                           source:  'string')
  end

  it 'does count all occurences' do
    src = <<-EOS
      class Outer
        # Method is necessary because we don't count empty classes.
        def something; end
        class Inner
        end
      end

    EOS

    expect(src).to reek_of(described_class,
                               lines:   [1],
                               context: 'Outer')
    expect(src).to reek_of(described_class,
                           lines:   [4],
                           context: 'Outer::Inner')
  end
end
```

The 2nd spec 'does count all occurences' is only added when it actually makes sense, so there are some detectors that only start with the first spec above.

* Removed contexts that didn't make any sense. E.g. I removed non-nested contexts with only example and just updated the example description. Same goes for specs that were
covered by one giant context.

mvz commented:
Awesome work, @troessner!

mvz commented on CONTRIBUTING.md:
```diff
@@ -52,12 +54,82 @@ start contributing.
+#### Spec naming conventions
+
+We do not use the popular "foo" / "bar" naming when it comes to the question "how to come up with good example names?".
+Instead, we use the [military alphabet](https://en.wikipedia.org/wiki/NATO_phonetic_alphabet) in descending order
```
`s/descending/ascending/`

troessner replied:
Errr, yeah, fixed!

mvz commented on spec/reek/smells/unused_private_method_spec.rb:
```diff
@@ -1,100 +1,147 @@
-          def start; end
-          def drive; end
+          def charlie; end
+          def charlie; end
```
Should be bravo, then charlie.

troessner replied:
Good catch, fixed!

troessner commented:
Status update:

* UtilityFunction + FeatureEnvy
* NestedIterators - RepeatedConditional

are still refactor worthy.
And then I need to go through each spec again once more.
And then it's reviewable ;)

waldyr commented:
I closed this by accident. I was going to comment something and then I decided not to comment. I ended up closing the issue. Really sorry.

troessner commented:
This bad boy is ready for review! The diff is probably wayyyyy to confusing, might make more sense to look at the specs directly.

mvz commented on CONTRIBUTING.md:
```diff
@@ -52,12 +54,82 @@ start contributing.
+
+All smell detector specs start out with 2 generic examples like this (the second one only if it makes sense):
+
+TODO: Example below does not get rendered properly. Wtf.
```
Seems to be fine now, so this line can go :tada:!

troessner replied:
Updated.

mvz commented on spec/reek/smells/irresponsible_module_spec.rb:
```diff
@@ -1,208 +1,191 @@
-    EOS
-    expect(src).to reek_of(:IrresponsibleModule, context: 'Foo')
-  end
+    it "does not report a namespace #{scope} that has a nested #{scope} through assignment" do
```
The nested item is always a class, so you shouldn't interpolate scope there.

troessner replied:
Good catch, thanks.

mvz commented on spec/reek/smells/nested_iterators_spec.rb:
```diff
@@ -100,152 +141,104 @@ def super_call_with_block
-          end
-        end
-      EOS
-      expect(source).to reek_of(:NestedIterators, lines: [3])
```
I think this test is still important, but this assertion can also be merged into the one that is now in line 128.

troessner replied:
Updated the example ending in line 128 to be

>>     expect(src).to reek_of(:NestedIterators, depth: 3, lines: [4])

mvz commented on spec/reek/smells/nested_iterators_spec.rb:
```diff
@@ -100,152 +141,104 @@ def super_call_with_block
-        end
-      EOS
-
-      expect(source).to reek_of(:NestedIterators, lines: [2, 3])
```
I think this case and the next are still important and not covered by the remaining tests. They are about the interaction of separate cases of NestedIterators within a single method.

troessner replied:
What do you mean? Isn't this case here covered by the "does count all occurrences" example in line 22 [here](https://github.com/troessner/reek/pull/1056/files#diff-b34733d371fc76dfe6eabf2134a73744R22)?

mvz replied:
No, that one has two cases with different nesting, and hence two smells, whereas this one has two cases with the same nesting, hence one smell with two lines.

troessner replied:
Good point, let me re-add this as separate commit so you can review it.

troessner replied:
@mvz added a spec for this as separate commit [here](https://github.com/troessner/reek/pull/1056/commits/44263ed14ff99dfc72f3fee05f861b0a2e37e0e0), please review ;)

mvz replied:
Looks good, :shipit:.

mvz commented on spec/reek/smells/smell_repository_spec.rb:
```diff
@@ -15,7 +15,7 @@
     end
 
-    it 'should return the smell types in alphabetic order' do
+    it 'should return the smell types in alfabetic order' do
```
-> alphabetic

mvz commented:
Phew, that **is** a big PR. I have put a few comments, but for the rest, I think we should merge this and then use mutant to sort out any bits we may have under-specified.

troessner commented:
@mvz addressed all of your feedback as far as possible. How about merging this one in its current state and then address whatever you feel missing in the NestedIterators specs afterwards?

mvz commented:
Ready to merge, @troessner!

troessner commented:
Squashed, please merge @mvz ;)

mvz commented:
:tada:
